code-me-a-hr.html

```diff
@@ -0,0 +1,856 @@
+	{
+	}
+
+	public function promote(string $toNewPosition)
+	{
+	}
+
+	public function fire(\DateTime $leftCompanyAt)
+	{
+	}
+
+	public function retire(\DateTime $leftCompanyAt)
+	{
+	}
+}
+						</code>
+					</pre>
+				</section>
+
+				<section>
+					<pre>
+						<code class="php hljs" data-trim >
+namespace Al\Domain;
+
+final class Employee implements EmployeeInterface
+{
+	/** @var Uuid */
+	private $id;
+
+	/** @var string */
+	private $name;
+
+	/** @var string */
+	private $position;
+
+	/** @var \DateTimeInterface */
+	private $hiredAt;
+
+	/** @var \DateTimeInterface */
+	private $firedAt = null;
+
+	/** @var \DateTimeInterface */
+	private $retiredAt = null;
+}
+						</code>
+					</pre>
+				</section>
+
+				<section>
+					<pre>
+						<code class="php hljs" data-trim >
+namespace Al\Domain;
+
+final class Employee implements EmployeeInterface
+{
+	// properties
+
+	private function __construct(Uuid $identifier, string $name, string $position, \DateTime $hiredAt)
+	{
+		$this->id = $identifier;
+		$this->name = $name;
+		$this->position = $position;
+		$this->hireAt = $hiredAt
+	}
+
+	public static function hire(Uuid $identifier, string $name, string $forPosition, \DateTime $hiredAt)
+	{
+		return new self($identifier, $name, $forPosition, $hiredAt);
+	}
+
+	public function promote(string $toNewPosition)
+	{
+		$this->position = $toNewPosition;
+	}
+
+	public function fire(\DateTime $leftCompanyAt)
+	{
+		$this->firedAt = $leftCompanyAt;
+	}
+
+	public function retire(\DateTime $leftCompanyAt)
+	{
+		$this->retiredAt = $leftCompanyAt;
+	}
+}
+						</code>
+					</pre>
+				</section>
+
+				<section>
+					<pre>
+						<code class="php hljs" data-trim >
+namespace Al\Domain;
+
+final class Employee implements EmployeeInterface
+{
+	private function __construct(Uuid $identifier, string $name, string $position, \DateTime $hiredAt)
+	{
+		if ($hiredAt < new \DateTime(2012-03-04)) { TODO: check that
+			throw new \OutOfRangeException('The company did not exist before 2012-03-04');
+		}
+
+		$this->id = $identifier;
+		$this->name = $name;
+		$this->position = $position;
+		$this->hireAt = $hiredAt
+	}
+
+	public function retire(\DateTime $leftCompanyAt)
+	{
+		if (null !== $this->fired) {
+			throw new \Exception(sprint('%s employee has been fired!', $this->name));
+		}
+
+		$this->retiredAt = $leftCompanyAt;
+	}
+}
+						</code>
+					</pre>
+				</section>
+
+				<section>
+					<h2>How can we use it in our application?</h2>
+				</section>
+
+				<section>
+					<h2><strong class="php-color">First problem:</strong> how do we use Doctrine?</h2>
+					<span class="fragment">
+						<pre>
+							<code class="php hljs" data-trim >
+								interface EmployeeRepository
+								{
+									public function findByNameAndPositionWithoutFiredPeople(string $name, string $position);
+								}
+							</code>
+						</pre>
+						<p>We use them as query repositories</p>
```

Comment:
dans le sens : on les utilise pour requeter les repos, tu devrais mettre "We use them **to** query repositories"